packages/dashboard/src/decorators/press.ts

```diff
@@ -0,0 +1,128 @@
+import { ComponentInterface } from '@stencil/core';
+import { noop } from 'lodash';
+import { Position } from '../types';
+import { distance } from '../util/distance';
+
+export type PressEvents = 'press';
+
+type PressTrigger = 'mouse' | 'touch' | 'all';
+export type PressOptions = {
+  duration?: number;
+  threshold?: number;
+  trigger?: PressTrigger;
+  target?: 'parent' | 'self';
+};
+
+const defaultOptions: Required<PressOptions> = {
+  duration: 500,
+  threshold: 5,
+  trigger: 'touch',
+  target: 'self',
+};
+
+type EventTypess = {
+  start: string;
+  move: string;
+  end: string;
+  cancel: string | null;
+};
+
+const getEvents = (trigger: PressTrigger): EventTypess => {
+  switch (trigger) {
+    case 'mouse':
+      return {
+        start: 'mousedown',
+        move: 'mousemove',
+        end: 'mouseup',
+        cancel: null,
+      };
+    case 'touch':
+      return {
+        start: 'touchstart',
+        move: 'touchmove',
+        end: 'touchend',
+        cancel: 'touchcancel',
+      };
+    case 'all':
+      return {
+        start: 'pointerdown',
+        move: 'pointermove',
+        end: 'pointerup',
+        cancel: 'pointercancel',
+      };
+  }
+};
+
+export const getEvent = (event: Event) =>
+  (event as TouchEvent).touches ? (event as TouchEvent).touches[0] : (event as MouseEvent);
+
+export const withPressHandler = (
+  component: ComponentInterface,
+  el: HTMLElement,
+  options: PressOptions,
+  cb: (event: Event) => void
+): (() => void) => {
+  let startPosition: Position | null = null;
+  let pressTimeout: NodeJS.Timeout;
+
+  const { duration, threshold, trigger, target } = Object.assign({}, defaultOptions, options);
+  const targetElement = target === 'parent' ? el.parentElement : el;
+
+  if (targetElement === null) return noop;
+
+  const { start, move, end, cancel } = getEvents(trigger);
+
+  const reset = () => {
+    startPosition = null;
+    clearTimeout(pressTimeout);
+  };
+
+  const pressStartHandler = (event: Event) => {
+    const ev = getEvent(event);
+    if (!ev) return;
+    startPosition = {
+      x: ev.clientX,
+      y: ev.clientY,
+    };
+    pressTimeout = setTimeout(() => {
+      cb.call(component, event);
+      reset();
+    }, duration);
+  };
+  const pressEndHandler = (event: Event) => {
+    if (startPosition === null || event.type === 'mouseup' || event.type === 'pointerup') {
```

Comment:
a common bug i see is where people handle null explicitly, and then eventually something changes and an `undefined` somehow gets passed down. we can just use `== null` and `!= null` instead and always just handle null and undefined - unless there is a really good reason otherwise.

Reply:
The typing on start position is defined as `let startPosition: Position | null = null;`. Unless someone overrides an assignment with an any or something similar this shouldn't be a problem? I can change the check to see if the variable is falsey to guard against that case though.